Mathlib/RingTheory/Flat/FaithfullyFlat.lean

```diff
@@ -0,0 +1,248 @@
+/-
+Copyright (c) 2024 Judith Ludwig, Florent Schaffhauser. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Judith Ludwig, Florent Schaffhauser.
+-/
+import Mathlib.RingTheory.Flat.Stability
+import Mathlib.RingTheory.IsTensorProduct
+import Mathlib.LinearAlgebra.TensorProduct.Tower
+import Mathlib.Algebra.Algebra.RestrictScalars
+import Mathlib.RingTheory.Finiteness
+import Mathlib.Algebra.Module.Defs
+
+/-!
+# Faithfully flat modules
+
+A module `M` over a commutative ring `R` is *faithfully flat* if it is flat and,
+for all `R`-module homomorphism `f : N → N'` such that `id ⊗ f = 0`, we have `f = 0`.
+
+In the stacks project, the definition of faithfully flat is different but tag
+<https://stacks.math.columbia.edu/tag/00TN> proves that their definition is equivalent to this.
+
+## Main declaration
+
+- `Module.FaithfullyFlat`: the predicate asserting that an `R`-module `M` is faithfully flat.
+
+## Main theorems
+
+- `Module.FaithfullyFlat.of_linearEquiv`: modules linearly equivalent to a flat modules are flat
+- `Module.FaithfullyFlat.comp`: https://stacks.math.columbia.edu/tag/00HC
+-/
+
+universe u v
+
+namespace Module
+
+variable (R : Type u) (M : Type v) [CommRing R] [AddCommGroup M] [Module R M]
+
+@[mk_iff] class FaithfullyFlat : Prop where
+  flat : Module.Flat R M := by infer_instance
+  zero_if_lTensor_zero :  ∀ ⦃N N': Type v⦄ [AddCommGroup N][Module R N][AddCommGroup N']
```

Comment:
The quantification over all `Type v` is not ideal. I'm not even sure if this implies faithfully flat for all `Type w`. Maybe `𝔪 • (⊤ : Submodule R M) ≠ ⊤` for all maximal ideal `𝔪` as the official definition?

Reply:
We've rewritten the file using just one universe level (the previous choice was due to the analogy with the case of [flat modules](https://github.com/leanprover-community/mathlib4/blob/master/Mathlib/RingTheory/Flat/Basic.lean)). We are working towards the proof of the equivalence between various definitions, but we'd like to keep this one as the main one for now, since we have already proven some properties using that characterisation of faithfully flat modules (WIP).

Reply:
For flat modules, the official definition is using ideals for this precise reason of avoiding universe issues. If we are sticking with this definition for now I prefer the hetero-universal one with the universe in the definition the same as the universe of the ring, which is at least the right definition. But please add the condition with general universe as a todo.